### DISCUSSION TITLE
Incompatible snowflake-connector-python & keyring versions in constraints file

### DISCUSSION THREAD

**Discussion by erikogan** [apache/airflow#41111] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.9.2

### What happened?

I am investigating the viability of MWAA for a project. I’m using [aws-mwaa-local-runner](https://github.com/aws/aws-mwaa-local-runner)‘s `mwaa-local-env test-requirements`. I’ve generated the `requirements.txt` from a Poetry lock file using the `poetry export` plugin.

However, the Airflow 2.9.2 / Python 3.11 [constraints file](https://raw.githubusercontent.com/apache/airflow/constraints-2.9.2/constraints-3.11.txt) specifies conflicting versions of some of my requirements:

* `snowflake-connector-python==3.10.1`, which requires:
   * ~[`urllib3` < 2.0.0](https://github.com/snowflakedb/snowflake-connector-python/blob/v3.10.1/setup.cfg#57)~
   * [`keyring` < 25.0](https://github.com/snowflakedb/snowflake-connector-python/blob/v3.10.1/setup.cfg#L99)
* `urllib3==2.2.1`
* `keyring==25.2.1`

(Updating to `snowflake-connector-python==3.12.0` would solve the `keyring` constraint, ~but it still requires `urllib3` < 2.0.0~)

I tried removing both `keyring` and `urllib3` from my `requirements.txt` after generation, which got me past the resolution while downloading dependencies, but the resolver still catches me during the actual install phase:

```
ERROR: Cannot install snowflake-connector-python[secure-local-storage]==3.10.1 because these package versions have conflicting dependencies.

The conflict is caused by:
    snowflake-connector-python[secure-local-storage] 3.10.1 depends on keyring<25.0.0 and >=23.1.0; extra == "secure-local-storage"
    The user requested (constraint) keyring==25.2.1

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict
```

**EDIT:**
I just realized `snowflake-connector-python` only requires `urllib3` when the Python version is less than 3.10. I was doing my `poetry export` in a legacy Python 3.9 environment. So this is a non-issue in Python 3.11, though `keyring` still is.

### What you think should happen instead?

The constraints file should not include conflicting versions of these modules, which prevents installation when `snowflake-connector-python` is required.

I believe the only way to fix this is to either downgrade the version of `keyring` to 24.x.y, or upgrade the version of `snowflake-connector-python` to 3.12.0. My preference would be for the latter, but either would unblock the process.

### How to reproduce

1. Clone https://github.com/aws/aws-mwaa-local-runner
2. Check out `v2.9.2` branch (currently the default)
3. Add the connector (with secure-local-storage, requiring keyring): `echo 'snowflake-connector-python[secure-local-storage]==3.10.1' >> requirements/requirements.txt`
4. Build the MWAA image: `./mwaa-local-env build-image`
5. Test the requirements: `./mwaa-local-env test-requirements`

**NOTE:** After that echo, the contents of the requirements file are:

```
--constraint "https://raw.githubusercontent.com/apache/airflow/constraints-2.9.2/constraints-3.11.txt"

apache-airflow-providers-snowflake==5.5.1
apache-airflow-providers-mysql==5.6.1
snowflake-connector-python[secure-local-storage]==3.10.1
```

### Operating System

Amazon Linux 2023.5.20240722 (In a Docker container)

### Versions of Apache Airflow Providers

In the reproduction case above, the following providers are included (beyond the base install)
```
apache-airflow-providers-snowflake==5.5.1
apache-airflow-providers-mysql==5.6.1
```

### Deployment

Amazon (AWS) MWAA

### Deployment details

I haven’t gotten to deployment yet, still trying to get it installed.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
See and read in detail the description of constraints. Constraints are only there to provide reproducible installation of Airlfow https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html#constraints-files - in EXACT set of  the "golden" set of requirements. If you want to upgrade any of those and change / upgrade any of the requirements of Airflow, you should have a second step of installation - where you diverge from the original constraints (and not use the constraints any more). 

Airflow deliberately has wide enough set of REQUIREMENTES that should allow you to diverge from the "golden set". But applying CONSTRAINTS (which has FIXED 719 or so depedencies) when at the same time you want to diverge from the constraints makes just no sense - you will always have conflicts. Constraints are not supposed to be use any more when you diverge from the "golden set".

This is explained in detail here https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html#constraints-files 

Also you can watch my talk from last Airflow summit describing the proces: https://airflowsummit.org/sessions/2023/mastering-dependencies-the-airflow-way/

Generallly - you get airlfow installed, then you update your own dependencies and then you can even generate your own constraints out of that when you arrive at good set of. 

Now - how this fits MWAA image and how constraints are used there, I have very little knowledge of - this is more of a question to MWAA team, but this is a general process that should be followed:

1) Install airflow with constraints to get reproducible installation
2) Modify the installation WITHOUT constraints when you want to change the dependencies
3) If you want to get it reproducible - freeze the resulting set of dependencies as "your own constraints".

Converting it into discussion in case more of the discussion is needed.

**Comment by potiuk** [comment]
BTW. Generally speaking - there is very little knowlede or help in this forum for MWAA-specific solutions/implementation - since MWAA is a paid service on top of Airflow, you are way better off asking your question to the support you are paying for - here people mostly help each other in their free time, as Airflow is Open-source project, created by volunteers.

**Comment by potiuk** [comment]
(and one more thing - generally if you have specific problem for an issue that does not seem like a generic Airflow issue - you are way better starting from creating a discussion rather than an issue) - issues are foreseen for case where it's quite probable that this is a generic issue related to Apache Airflow (vanilla, open-source) and when ideally there is a reproducible case that allows to diagnose and fix the issue - other than that, most of such things are at most discussions - where people might get help from others.

**Comment by vtrbraga** [comment]
Having the same issue in our deployment.

@potiuk the issue here is that we're not trying to diverge from the "golden set" of constraints, but rather that the constraint golden set contains a version mismatch between two packages, where the pinned `snowflake-connector-python==3.10.1` requires `keyring<25.0.0`, but the constraints also pin `keyring==25.2.1`, which makes it impossible to install snowflake-connector-python using this constraints file, even when not diverging from specifications.

**Comment by potiuk** [comment]
The `keyring` is not a requirement of snowflake-connection-python. It's a requirement of it's "secure-local-storage" extra which is not installed as part of our "golden set". 

It's simply impossible - with our 719 dependencies to make sure that the golden set contains all the combinations of all the extras from all the dependencies. 

In this case this is exactly what I described - by installing [secure-local-storage] you are diverging from the golden set - so it should be done as a separate step without constratins - and let `pip` find the right set of non-conflicting dependencies.

Snowflake connector-python metadata:

```
Requires-Dist: keyring <25.0.0,>=23.1.0 ; extra == 'secure-local-storage'
```

Snowflake provider dependencies (as you see `secure-local-storage`  is not there).

https://github.com/apache/airflow/blob/main/airflow/providers/snowflake/provider.yaml#L77

```
dependencies:
  - apache-airflow>=2.7.0
  - apache-airflow-providers-common-compat>=1.1.0
  - apache-airflow-providers-common-sql>=1.14.1
  # In pandas 2.2 minimal version of the sqlalchemy is 2.0
  # https://pandas.pydata.org/docs/whatsnew/v2.2.0.html#increased-minimum-versions-for-dependencies
  # However Airflow not fully supports it yet: https://github.com/apache/airflow/issues/28723
  # In addition FAB also limit sqlalchemy to < 2.0
  - pandas>=2.1.2,<2.2;python_version>="3.9"
  - pandas>=1.5.3,<2.2;python_version<"3.9"
  - pyarrow>=14.0.1
  - snowflake-connector-python>=3.7.1
  - snowflake-sqlalchemy>=1.4.0
```

**Comment by jbernardez-om** [comment]
Hi, I am experiencing the same issue trying to get the following requirements in my MWAA 2.9.2 setup :
dbt-core==1.8.0
dbt-snowflake==1.8.0

The conflict is caused by:
    snowflake-connector-python[secure-local-storage] 3.10.1 depends on keyring<25.0.0 and >=23.1.0; extra == "secure-local-storage"
    The user requested (constraint) keyring==25.2.1
    
Is there any update on this issue ? I understand it is most likely MWAA related but was just wondering if someone managed to get this sorted ?

**Reply by potiuk** [reply]
You can create your own constraints file or install packages without constraints - not sure how any of those options is available in MWAA though.

**Reply by jbernardez-om** [reply]
I found out how to fix it. I took the official constraints file for version 2.9.2 and removed keyring. Uploaded the constraints file to my s3 bucket/dags folder and then pointed the requirements.txt file to my own constraints file using the following :

--constraint "/usr/local/airflow/dags/constraints.txt"
